hdf5-rs/src/hl/container.rs

```diff
@@ -35,19 +36,56 @@ impl<'a> Reader<'a> {
         self
     }
 
-    fn read_into_buf<T: H5Type>(&self, buf: *mut T) -> Result<()> {
+    fn read_into_buf<T: H5Type>(&self, buf: *mut T, file_space: Option<(&Dataspace, &Dataspace)>) -> Result<()> {
+
         let file_dtype = self.obj.dtype()?;
         let mem_dtype = Datatype::from_type::<T>()?;
         file_dtype.ensure_convertible(&mem_dtype, self.conv)?;
         let (obj_id, tp_id) = (self.obj.id(), mem_dtype.id());
+        let (fspace_id, mspace_id) = file_space.map(|(f, m)| (f.handle().id(), m.handle().id())).unwrap_or((H5S_ALL, H5S_ALL));
+
         if self.obj.is_attr() {
             h5try!(H5Aread(obj_id, tp_id, buf as *mut _));
         } else {
-            h5try!(H5Dread(obj_id, tp_id, H5S_ALL, H5S_ALL, H5P_DEFAULT, buf as *mut _));
+            h5try!(H5Dread(obj_id, tp_id, mspace_id, fspace_id, H5P_DEFAULT, buf as *mut _));
         }
         Ok(())
     }
 
+    /// Reads a slice of an n-dimensional array.
+    /// If the array has a fixed number of dimensions, it must match the dimensionality of
+    /// dataset.  FIXME - ensure this: Sliced reads are not allowed from attributes.
+    /// Use the multidimsional slice macro `s![]` from `ndarray` to conveniently create
+    /// a multidimensional slice.
+    pub fn read_slice<T,S,D>(&self, slice: SliceInfo<S,D>) -> Result<Array<T, D>>
```

Comment:
Shouldn't the slice be a ref?

Looking at [`ArrayBase::slice()`](https://docs.rs/ndarray/0.12.1/ndarray/struct.ArrayBase.html#method.slice), it takes a `&SliceInfo<...>`?